### ISSUE TITLE
storage: unexpected slow intentInterleavingIter seek

### ISSUE THREAD

**Issue by jbowens** [cockroachdb/cockroach#89327]
https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1664898054194479

<img width="977" alt="image" src="https://user-images.githubusercontent.com/867352/193916563-a960b106-dc48-491b-87b8-5ea8842335f5.png">

This profile is unexpected. Something seems either wrong or there's a significant missed optimization here.

Jira issue: CRDB-20205

**Comment by sumeerbhola** [comment]
We have a working theory based on the understanding that many range tombstones were being added to the TPCC ranges while the KV workload was running (the latter is experiencing these slow MVCCGets). Details in the above slack thread, and copying the summary here:

The prefix seekGE for this engineIter has no limit key and no upper bound (from the caller intentInterleavingIter). pebble.Iterator.SeekPrefixGE does not even support a WithLimit variant because the assumption was that some part of the iterator tree would quickly stop when past the prefix (so a WithLimit variant is unnecessary). Our current early stopping logic for prefix iteration is based on one of (a) the sstable iterator fails the bloom filter match, and then we rely on levelIter translating this into something that the mergingIter will not advance past to load the next file, (b) mergingIter notices that it has advanced past the prefix, (b) the top-level Iterator notices it has advanced past the prefix.
(a) is not happening here because of not using bloom filters (even if it were happening, we can get unlucky for some queries). (b) is gated on the reseeked bool in mergingIter and in our case we are not seeking -- the mergingIter is encountering a lot of range tombstones in the lock table key space corresponding to ranges that were removed (in the TPCC workload, and not part of this KV range, but since we don't set an upper bound the iterator doesn't know). (c) is too late.

One fix would be to make mergingIter.findNextEntry be more aggressive about checking that we are not past the prefix. We don't want mergingIter to check keys that it is returning to the top-level Iterator since that would duplicate work that the latter is already doing. But maybe we should make it check everytime it loops in findNextEntry or do it every k iterations as mentioned in the existing code comment.
